File: Android/app/src/main/java/com/kite/zmusic/data/LrcParser.kt
package com.kite.zmusic.data

/**
 * 网易云 [mm:ss.xx] / [mm:ss] 行歌词解析。
 */
data class LrcLine(
    val timeMs: Long,
    val text: String,
)

object LrcParser {

    private val lineRegex = Regex("""^\[(\d{1,2}):(\d{2})(?:\.(\d{1,3}))?](.*)$""")

    fun parse(raw: String): List<LrcLine> {
        val out = ArrayList<LrcLine>()
        for (line in raw.lineSequence()) {
            val t = line.trim()
            if (t.isEmpty()) continue
            val m = lineRegex.find(t) ?: continue
            val mm = m.groupValues[1].toLongOrNull() ?: continue
            val ss = m.groupValues[2].toLongOrNull() ?: continue
            val frac = m.groupValues[3]
            val subMs = parseFractionMs(frac)
            val text = sanitizeLyricText(m.groupValues[4]) ?: continue
            val base = (mm * 60L + ss) * 1000L + subMs
            out.add(LrcLine(base, text))
        }
        out.sortBy { it.timeMs }
        return out
    }

    /**
     * 去掉空白/无意义占位，获取后即净化，UI 不再出现空行。
     * @return null 表示应丢弃该行
     */
    fun sanitizeLyricText(raw: String): String? {
        val t = raw.trim()
            .replace('\u00A0', ' ')
            .replace(Regex("\\s+"), " ")
            .trim()
        if (t.isEmpty()) return null
        // 纯标点 / 占位符不算有效歌词
        if (t.all { it.isWhitespace() || it in "·.•…-_—~/|" }) return null
        return t
    }

    private fun parseFractionMs(frac: String): Long {
        if (frac.isEmpty()) return 0L
        return when (frac.length) {
            1 -> frac.toLongOrNull()?.times(100L) ?: 0L
            2 -> frac.toLongOrNull()?.times(10L) ?: 0L
            else -> frac.take(3).toLongOrNull()?.coerceAtMost(999L) ?: 0L
        }
    }
}

File: Windows/Data/LrcParser.cs
using System.Text.RegularExpressions;

namespace ZMusic.Data;

/// <summary>
/// Netease-style [mm:ss.xx] / [mm:ss] LRC parser (Windows port of Android LrcParser).
/// </summary>
public static partial class LrcParser
{
    [GeneratedRegex(@"^\[(\d{1,2}):(\d{2})(?:\.(\d{1,3}))?](.*)$", RegexOptions.Compiled)]
    private static partial Regex LineRegex();

    public static IReadOnlyList<LrcLine> Parse(string raw)
    {
        var outList = new List<LrcLine>();
        foreach (var line in raw.Split(['\r', '\n'], StringSplitOptions.RemoveEmptyEntries))
        {
            var t = line.Trim();
            if (t.Length == 0)
            {
                continue;
            }

            var m = LineRegex().Match(t);
            if (!m.Success)
            {
                continue;
            }

            if (!long.TryParse(m.Groups[1].Value, out var mm) ||
                !long.TryParse(m.Groups[2].Value, out var ss))
            {
                continue;
            }

            var text = SanitizeLyricText(m.Groups[4].Value);
            if (text is null)
            {
                continue;
            }

            var subMs = ParseFractionMs(m.Groups[3].Value);
            outList.Add(new LrcLine((mm * 60L + ss) * 1000L + subMs, text));
        }

        outList.Sort((a, b) => a.TimeMs.CompareTo(b.TimeMs));
        return outList;
    }

    public static string? SanitizeLyricText(string raw)
    {
        var t = raw.Trim()
            .Replace('\u00A0', ' ');
        t = WhitespaceRegex().Replace(t, " ").Trim();
        if (t.Length == 0)
        {
            return null;
        }

        if (t.All(ch => char.IsWhiteSpace(ch) || "·.•…-_—~/|".Contains(ch)))
        {
            return null;
        }

        return t;
    }

    private static long ParseFractionMs(string frac)
    {
        if (string.IsNullOrEmpty(frac))
        {
            return 0;
        }

        return frac.Length switch
        {
            1 => long.TryParse(frac, out var a) ? a * 100L : 0L,
            2 => long.TryParse(frac, out var b) ? b * 10L : 0L,
            _ => long.TryParse(frac.Length > 3 ? frac[..3] : frac, out var c)
                ? Math.Min(c, 999L)
                : 0L,
        };
    }

    [GeneratedRegex(@"\s+", RegexOptions.Compiled)]
    private static partial Regex WhitespaceRegex();
}
